### PR TITLE
refactor: Refactor `MaterialInteractor`

### PR DESCRIPTION
**Pull request by andiwand**
Aimed to improve consistency and split surface and volume interactions into two logical parts

pulled out of https://github.com/acts-project/acts/pull/2603

### CODE REVIEW

File: Core/include/Acts/Propagator/MaterialInteractor.hpp [hunk 1]
```diff
@@ -55,74 +56,85 @@ struct MaterialInteractor {
...
+    // interaction code
+    const Surface* surface = navigator.currentSurface(state.navigation);
+
+    // We only have material interactions if there is potential material
     if (surface && surface->surfaceMaterial()) {
+      ACTS_VERBOSE("MaterialInteractor | "
+                   << "Found material on surface " << surface->geometryId());
+
       // Prepare relevant input particle properties
       detail::PointwiseMaterialInteraction d(surface, state, stepper);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MaterialInteractor.hpp:80]: Not part of the change set, but `d` really is that great of a variable name for this.

File: Core/include/Acts/Propagator/MaterialInteractor.hpp [hunk 2]
```diff
@@ -159,16 +171,18 @@ struct MaterialInteractor {
...
-    mi.position = d.pos;
-    mi.time = d.time;
-    mi.direction = d.dir;
-    mi.surface = nullptr;
-    mi.volume = d.volume;
-    mi.pathCorrection = d.pathCorrection;
-    mi.materialSlab = d.slab;
-    result.materialInteractions.push_back(std::move(mi));
+    // Record the interaction if requested
+    if (recordInteractions) {
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MaterialInteractor.hpp:175]: Not important but could be an early return instead.

File: Core/include/Acts/Propagator/MaterialInteractor.hpp [hunk 3]
```diff
@@ -179,19 +193,21 @@ struct MaterialInteractor {
...
-        momentum - result.materialInteractions.back().direction.norm();
-    result.materialInteractions.back().materialSlab.scaleThickness(
-        shift.norm());
-    result.materialInteractions.back().updatedVolumeStep = true;
-    result.materialInX0 +=
-        result.materialInteractions.back().materialSlab.thicknessInX0();
-    result.materialInL0 +=
-        result.materialInteractions.back().materialSlab.thicknessInL0();
+    // Update the previous interaction if requested
+    if (recordInteractions) {
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MaterialInteractor.hpp:197]: Same here.

**Review by paulgessinger — DISMISSED** [review]
Nothing substantial from my side. Seems like a good idea.

**Review by paulgessinger — DISMISSED** [review]

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [825db48eaffae30bcba8c6eee1a9c3a37a09519e]
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100657) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100656) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100655) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100654) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100653) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100652) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100651) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100650) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100649) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100648) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100647) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100646) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100645) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100644) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100643) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100642) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100641) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100640) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100639) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34100638) |
